src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -186,22 +196,108 @@ public void close() throws IOException
       {
         open = false;
         internalFiles.put(name, new Metadata(currOut.getFileNum(), startOffset, currOut.getCurrOffset()));
+        writerCurrentlyInUse = false;
 
         if (bytesWritten != currOut.getCurrOffset() - startOffset) {
           throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
         }
         if (bytesWritten != size) {
           throw new IOException(
               String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
-          );
+              );
+        }
+
+        if(!writerCurrentlyInUse)
+        {
+          mergeWithSmoosher();
+        }
+      }
+    };
+  }
+
+  private void mergeWithSmoosher() throws IOException
+  {
+    //get processed elements from the stack and write.
+    List<File> fileToProcess = new ArrayList<>(files);
+    files = Lists.newArrayList();
+    for(File file: fileToProcess)
+    {
+      add(file);
+      createFiles.remove(file);
+      file.delete();
+    }
+  }
+
+  private SmooshedWriter delegateSmooshedWriter(final String name,final long size) throws IOException
+  {	
+    final File tmpFile = new File(baseDir, name);
+    createFiles.add(tmpFile);
+    return new SmooshedWriter () {
+      private int currOffset = 0;
+      private boolean open = true;
+      private final FileOutputStream out = new FileOutputStream(tmpFile);
+
+      @Override
+      public void close() throws IOException
+      {
+        open = false;
+        out.close();
+        files.add(tmpFile);
```

Comment:
Why wouldn't we remove the file from the `createFiles` or the `filesInProcess` here?